wallet/src/account/output_cache/mod.rs

```diff
@@ -732,59 +745,143 @@ impl OutputCache {
+                        }),
+                        | AccountCommand::FreezeToken(token_id, _) => Some(ConflictCheck {
+                            frozen_token_id: Some(*token_id),
+                            confirmed_account_nonce: Some((cmd.into(), *nonce)),
+                        }),
+                    },
+                }
+            })
+            .collect::<Vec<_>>();
+
+        // Collect all conflicting txs
+        let mut conflicting_txs = BTreeSet::new();
 
-        let mut conflicting_txs = vec![];
-        if let Some(frozen_token_id) = frozen_token_id {
+        for conflict_check in conflict_checks {
             for unconfirmed in self.unconfirmed_descendants.keys() {
                 let unconfirmed_tx = self.txs.get(unconfirmed).expect("must be present");
-                if self.uses_token(unconfirmed_tx, frozen_token_id) {
-                    let unconfirmed_tx = self.txs.get_mut(unconfirmed).expect("must be present");
-                    match unconfirmed_tx {
-                        WalletTx::Tx(ref mut tx) => {
-                            tx.set_state(TxState::Conflicted(block_id));
+
+                match unconfirmed_tx {
+                    WalletTx::Tx(tx) => {
+                        if let Some(frozen_token_id) = conflict_check.frozen_token_id {
+                            if self.violates_frozen_token(unconfirmed_tx, &frozen_token_id) {
+                                conflicting_txs.insert(tx.get_transaction().get_id());
+                                continue;
+                            }
                         }
-                        WalletTx::Block(_) => {}
-                    };
 
-                    conflicting_txs.push(unconfirmed);
-                }
+                        if let Some((confirmed_account, confirmed_account_nonce)) =
+                            conflict_check.confirmed_account_nonce
+                        {
+                            if confirmed_tx.get_id() != tx.get_transaction().get_id()
+                                && uses_conflicting_nonce(
+                                    unconfirmed_tx,
+                                    confirmed_account,
+                                    confirmed_account_nonce,
+                                )
+                            {
+                                conflicting_txs.insert(tx.get_transaction().get_id());
+                                continue;
+                            }
+                        }
+                    }
+                    WalletTx::Block(_) => {
+                        utils::debug_panic_or_log!("Cannot be block reward");
+                    }
+                };
+            }
+        }
+
+        // Remove all descendants of conflicting txs
+        let mut conflicting_tx_ids_with_descendants = vec![];
+        let mut conflicting_txs_with_descendants = vec![];
+
+        for conflicting_tx in conflicting_txs {
```

Comment:
[discussed] So, there are still problems here - if in `conflicting_txs` a parent comes before a child, on the 2nd iteration of the loop we'll get a tx that is already marked as conflicted, so we'll hit the `CannotMarkTxAsConflictedIfInState` error. The main question was why the existing tests didn't catch this.

Reply:
Fixed, test added